cmd/manufacturing.go

```diff
@@ -0,0 +1,330 @@
+var manufacturingCmd = &cobra.Command{
+	Use:   "manufacturing http_address",
+	Short: "Serve an instance of the HTTP server for the role",
+	Long: `Serve runs the HTTP server for the FDO protocol. It can act as all the three
+	main servers in the FDO spec.`,
+	Args: func(cmd *cobra.Command, args []string) error {
+		if err := cobra.ExactArgs(1)(cmd, args); err != nil {
+			return err
+		}
+		address = args[0]
+		return nil
+	},
+	RunE: func(cmd *cobra.Command, args []string) error {
+		state, err := getState()
+		if err != nil {
+			return err
+		}
+
+		err = db.InitDb(state)
+		if err != nil {
+			return err
+		}
+
+		// Retrieve RV info from DB
+		rvInfo, err := rvinfo.FetchRvInfo()
+		if err != nil {
+			return err
+		}
+
+		return serveManufacturing(rvInfo, state, insecureTLS)
+	},
+}
+
+// Server represents the HTTP server
+type ManufacturingServer struct {
+	addr    string
+	extAddr string
+	handler http.Handler
+	useTLS  bool
+}
+
+// NewServer creates a new Server
+func NewManufacturingServer(addr string, extAddr string, handler http.Handler, useTLS bool) *ManufacturingServer {
+	return &ManufacturingServer{addr: addr, extAddr: extAddr, handler: handler, useTLS: useTLS}
+}
+
+// Start starts the HTTP server
+func (s *ManufacturingServer) Start() error {
+	srv := &http.Server{
+		Handler:           s.handler,
+		ReadHeaderTimeout: 3 * time.Second,
+	}
+
+	// Channel to listen for interrupt or terminate signals
+	stop := make(chan os.Signal, 1)
+	signal.Notify(stop, syscall.SIGINT, syscall.SIGTERM)
+
+	// Goroutine to listen for signals and gracefully shut down the server
+	go func() {
+		<-stop
+		slog.Debug("Shutting down server...")
+
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+
+		if err := srv.Shutdown(ctx); err != nil {
+			slog.Debug("Server forced to shutdown:", "err", err)
+		}
+	}()
+
+	// Listen and serve
+	lis, err := net.Listen("tcp", s.addr)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = lis.Close() }()
+	slog.Info("Listening", "local", lis.Addr().String(), "external", s.extAddr)
+
+	if s.useTLS {
+		preferredCipherSuites := []uint16{
+			tls.TLS_AES_256_GCM_SHA384,                  // TLS v1.3
+			tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,   // TLS v1.2
+			tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, // TLS v1.2
+			tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, // TLS v1.2
+		}
+
+		if serverCertPath != "" && serverKeyPath != "" {
+			srv.TLSConfig = &tls.Config{
+				MinVersion:   tls.VersionTLS12,
+				CipherSuites: preferredCipherSuites,
+			}
+			return srv.ServeTLS(lis, serverCertPath, serverKeyPath)
+		} else {
+			return fmt.Errorf("no TLS cert or key provided")
+		}
+	}
+	return srv.Serve(lis)
+}
+
+func serveManufacturing(rvInfo [][]protocol.RvInstruction, db *sqlite.DB, useTLS bool) error {
+	key, keyType, err := parsePrivateKey(manufacturingKey)
+	if err != nil {
+		return err
+	}
+	c, err := os.ReadFile(deviceCert)
```

Comment:
Not a comment but for my own information: what is the use-case where the device x509 change needs to be provided? Would a device-specific cert need to be supplied for each device this server DI's?